GameOfRunes/GameOfRunes/GameEngine/GameEngine.swift

```diff
@@ -145,61 +155,111 @@ class GameEngine {
         guard let playerManaEntity = playerManaEntity else {
             return
         }
         
         systemDelegate.increaseMana(by: manaPoints, for: playerManaEntity)
     }
-
+    
     func decreasePlayerMana(by manaPoints: Int) {
         increasePlayerMana(by: -manaPoints)
     }
+}
 
-    func didActivatePowerUp(powerUp: PowerUpType, at position: CGPoint, with size: CGSize = .zero) -> Bool {
+/** Extension to the GameEngine for PowerUps */
+extension GameEngine {
+    func didActivatePowerUp(at position: CGPoint, size: CGFloat? = nil) -> Bool {
         guard let gameScene = gameScene,
             let playerManaEntity = playerManaEntity,
-            let currentManaPoints = systemDelegate.getMana(for: playerManaEntity) else {
+            let currentManaPoints = systemDelegate.getMana(for: playerManaEntity),
+            let selectedPowerUp = gameScene.powerUp() else {
                 fatalError("Invalid call to didActivatePowerUp")
         }
-                                                        // TODO: change this once game meta-data is up
-        let manaPointsRequired = powerUp.manaUnitCost * gameScene.playerAreaNode.manaBarNode.manaPointsPerUnit
         
+        // TODO: change this once game meta-data is up
+        let manaPointsRequired = selectedPowerUp.manaUnitCost * gameScene.playerAreaNode.manaBarNode.manaPointsPerUnit
+        gameScene.deselectPowerUp()
         guard currentManaPoints >= manaPointsRequired else {
             // did not activate
             return false
         }
         
-        switch powerUp {
+        var powerUpEntity: Entity
+        
+        switch selectedPowerUp {
         case .darkVortex:
             let radius = gameScene.size.width / 3
-            let powerUpEntity = DarkVortexPowerUpEntity(
+            powerUpEntity = DarkVortexPowerUpEntity(
                 gameEngine: self,
                 at: position,
                 with: .init(width: radius, height: radius)
             )
-            decreasePlayerMana(by: manaPointsRequired)
             add(powerUpEntity)
         case .hellfire:
-            print("Not Implemented Yet")
+            powerUpEntity = HellfirePowerUpEntity(
+                gameEngine: self,
+                at: position,
+                with: .init(width: (size ?? 0) * 2, height: (size ?? 0) * 2)
+            )
+            add(powerUpEntity)
         case .icePrison:
-            print("Not Implemented Yet")
+            powerUpEntity = IcePrisonPowerUpEntity(
+                gameEngine: self,
+                at: position,
+                with: .init(width: (size ?? 0) * 2, height: (size ?? 0) * 2)
+            )
+            add(powerUpEntity)
+            activateIcePrison()
         }
+        decreasePlayerMana(by: manaPointsRequired)
         
         // did activate
         return true
     }
+    
+    private func activateIcePrison() {
+        guard let icePrisonComponent = entities(for: .player)
```

Comment:
use `entities[.icePrisonEntity].first as? IcePrisonPowerUpEntity` instead since we are in the game engine

Reply:
or alternatively use `entities(for: .icePrisonEntity)`